### DISCUSSION TITLE
Proposal: Automated Calibration and Scenario Discovery

### DISCUSSION THREAD

**Discussion by liyuezha** [mesa/mesa#3207] [Ideas]
Building on the discussion in [#3176](https://github.com/mesa/mesa/issues/3176), which outlines the structure and intent of `ScenarioSpace`, I wanted to explore a possible extension focused on scenario discovery.

As `ScenarioSpace` grows, it can become challenging for users to manually explore large parameter spaces and identify scenarios that exhibit critical or interesting behaviors (e.g., regime shifts, or rare outcomes). Rather than treating `ScenarioSpace` solely as a passive collection of predefined scenarios, one could also view it as a searchable domain for systematic exploration.

**Idea: Search-based Discovery and Calibration**:
The idea is to introduce an optional search interface that operates on top of `ScenarioSpace`. This layer would allow users to programmatically discover scenarios that best matches user-defined metrics.

Key Components:
- Searchable Parameter Mapping: A mechanism to transform `ScenarioSpace` constraints into a format compatible with optimization backends (e.g., Optuna, Scipy.optimize).

- Metric-Driven Search: Users define a Metric (e.g., agent state proportions, spatial patterns), , and the search process explores the space to surface scenarios of interest.

**GSoC 2026 Aspiration:** I would love to hear whether a search-based discovery layer would be useful to Mesa users. Would the community be interested in a GSoC project focusing on this?

**Comment by quaquel** [comment]
There are already other Python libraries that cover this type of functionality, and they are easy to integrate with Mesa (e.g., ema_workbench). There is also an entire scientific field working on this topic. So, I am not sure this is the ideal focus for a Mesa GSoC project.

**Reply by liyuezha** [reply]
Thank you for your thoughtful feedback！I completely agree that this is an established research area and that Mesa should not aim to re-implement functionality already well covered by existing libraries.

To clarify, my intention is not to replicate optimization algorithms themselves. Rather, the idea is to explore whether Mesa could benefit from a lightweight, standardized interface that makes it easier to connect Scenario with external search or discovery tools.

One motivation for this is that Mesa models often produce structured outputs (e.g., grid-based spatial states, agent distributions) that require model-specific glue code before they can be evaluated by external tooling. In practice, users frequently need to write custom adapters to translate Mesa outputs into metrics consumable by general-purpose libraries.

A possible direction (purely as an example) could be to provide a small set of optional, Mesa-native metric helpers (e.g., spatial overlap measures on grids) that make it easier to expose such outputs in a consistent way, while leaving the actual search, optimization, and experimental design fully to external libraries.

I fully agree that this should remain an optional integration pattern rather than a core Mesa feature. If the scope were narrowed to focus specifically on spatial validation metrics and a lightweight external connector interface, would this be considered a appropriately scoped direction to explore for a GSoC project?

**Reply by EwoutH** [reply]
I think there might be some value in automatic calibration. However, this field comes with many challenges, especially in stochastic ABMs. For example how do you calibrate high dimensional ABMs, deal with colinarity between variables and prevent overfitting?

I see the final product as:
- A thorough analysis of the state of the field
- An analysis of do and don’t for Mesa (based on experimentation insights)
- Documentation like example models and tutorials

**Comment by wang-boyu** [comment]
Thanks for proposing the idea. I think it can be quite useful, since model calibration is often viewed as a standard step in typical ABM workflows. As mentioned above, there're existing calibration (optimization) frameworks available. So I imagine the focus of this proposal would be **API design** and **integration**, as you commented above:

1. design a set of standard calibration, or optimization, or parameter search/exploration, APIs and data formats in Mesa. Likely build on to of Mesa's existing BatchRunner and Scenario.
2. interface to 3rd party optimizers, or implement relatively straightforward algorithms as examples, so that users can implement their own by following them
3. provide supporting tools such as visualization, progress monitoring, and logging components
4. documentation, example workflow, test, etc

> For example how do you calibrate high dimensional ABMs, deal with colinarity between variables and prevent overfitting?

This is a valid concern and a real issue, but I don't really think it's the problem we're trying to solve here. We provide tools (or integrate other tools) and let users to decide which one to use on their own discretion. Maybe they come up with their own optimization algorithms and plug into our calibration API, while re-using other modules such as BatchRunner or visualization tools.

For comparison here are some similar functionalities in other ABM frameworks. Some choose to integreate, and some others implement their own optimizers:

1. [`Calibrator`](https://abm4all.github.io/Melodie/html/api/calibrator.html) in [Melodie](https://abm4all.github.io/Melodie/html/index.html), another Python ABM framework
2. [BehaviorSearch](https://www.behaviorsearch.org/) for NetLogo
3. [Calibration experiment](https://anylogic.help/anylogic/experiments/calibration.html) framework in AnyLogic
4. [Exploration calibration methods](https://gama-platform.org/wiki/ExplorationMethods) in GAMA

**Reply by quaquel** [reply]
> This is a valid concern and a real issue, but I don't really think it's the problem we're trying to solve here. We provide tools (or integrate other tools) and let users to decide which one to use on their own discretion. Maybe they come up with their own optimization algorithms and plug into our calibration API,

Well said! In the [Mesa 4 discussion](https://github.com/mesa/mesa/discussions/2972#discussioncomment-15427010), I outlined my vision on the future of the experimentation API. This is a broad sketch, and we are starting to fill in parts of this with the new `Scenario` class and the ongoing work on `DataSet` and a reactive `Collector` API. 

It might be worthwhile to analyse how other ABM software, and perhaps even broader, devs software, handle this. Distil this into key components and assess how Mesa's API can be improved/refined. Resulting in a proof of concept where Mesa, with a potentially updated/improved experimentation api, is connected to several existing Python libraries for computational experimentation, optimization-based calibration, and Markov chain Monte Carlo (which is, in my view, the proper way to calibrate ABMs).

**Comment by liyuezha** [comment]
Thank you all for the feedback! I really appreciate the insights @EwoutH, @wang-boyu, and @quaquel.

I completely agree that the focus should be on **API design and integration** rather than solving the inherent complexities of calibration itself. The goal would be to provide a flexible framework that lets users choose their own optimization strategies while benefiting from Mesa's existing infrastructure.

## Framework Servey Summary

Following your suggestions, I've conducted a preliminary survey of calibration approaches in other ABM frameworks. Here's what I found:

### 1. **Melodie** (Python)
Melodie provides a dedicated `Calibrator` module as a "modeling manager" component that sits above the model level:
- **Architecture**: Separates calibration as a distinct component from the model itself
- **Objective**: Minimizes distance between model output and empirical data/target
- **Method**: Uses evolutionary training and optimization approaches
- **Integration**: Works with their `Scenario` and `DataCollector` components
- **Example**: [CovidContagionCalibrator](https://github.com/ABM4ALL/Melodie/tree/master/examples/covid_contagion_calibrator) demonstrates the implementation

**Key Takeaway**: The separation of calibration as a standalone "manager" that orchestrates model runs is a clean architectural pattern.

### 2. **BehaviorSearch** (NetLogo)
BehaviorSearch is a standalone tool bundled with NetLogo 6+:
- **Architecture**: External tool that interfaces with NetLogo via defined experiment protocols
- **Methods**: Genetic algorithms, simulated annealing, hill climbing, and other heuristic search methods
- **Workflow**: Users define parameter ranges, objective measures, and search algorithms through a GUI or config files
- **Complementary Tools**: Works alongside RNetLogo for more complex calibrations

**Key Takeaway**: Calibration is treated as an experiment design problem and  interact with models as black-box evaluators. Objective is a function over model outputs, not a property of the model itself.

### 3. **AnyLogic**
AnyLogic has a built-in `Calibration Experiment` framework:
- **Architecture**: Calibration is a first-class experiment type alongside Simulation, Optimization, and Monte Carlo experiments
- **Engines**: Supports both Genetic Algorithm and OptQuest (proprietary optimizer)
- **Features**: 
  - Define varied parameters with min/max ranges
  - Multiple calibration criteria with coefficients
  - Built-in `difference()` functions for comparing datasets
  - Progress visualization and automatic stopping conditions
- **Integration**: Calibrated parameter values can be easily transferred to simulation experiments

**Key Takeaway**: Treating calibration as a distinct experiment type with standardized APIs for objectives, constraints, and stopping conditions provides clarity.

### 4. **GAMA Platform**
GAMA distinguishes between exploration and calibration in batch experiments:
- **Exploration Methods**: Factorial, Latin Hypercube, Saltelli, Morris, Orthogonal, Uniform sampling
- **Calibration Methods**: Hill Climbing, Simulated Annealing, Tabu Search, Reactive Tabu, Genetic Algorithm, PSO
- **Architecture**: Uses a `method` facet in batch experiments with `minimize` or `maximize` objectives
- **Flexibility**: Allows custom aggregation functions (min, max, average) across replications
- **Server Mode**: Calibration methods are accessible via server API for R/Python integration

**Key Takeaway**: Clear separation between parameter space exploration (understanding model behavior) and calibration (matching target outputs) helps users choose appropriate methods.

### 5. **Conclusion**

Despite treating calibration either as an object or as a specialized experiment type, above systems share several core assumptions:

- 1.Parameters are explicitly declared, including their domains, types, and constraints.
- 2.Models are treated as black-box evaluators, decoupled from optimization or inference logic.
- 3.Calibration operates at the experiment level, not within the model definition.
- 4.Results are structured objects, often extending beyond a single “best” parameter set to include trade-offs, distributions, or response surfaces.

## References
- Melodie Framework: https://github.com/ABM4ALL/Melodie
- BehaviorSearch: https://www.behaviorsearch.org/
- AnyLogic Calibration: https://anylogic.help/anylogic/experiments/calibration.html
- GAMA Exploration Methods: https://gama-platform.org/wiki/ExplorationMethods

**Comment by liyuezha** [comment]
## Proposed Direction for Mesa

Based on this research and your previous feedback, I'm refining the proposal to emphasize **Experiment API**:

### 1. **Unified Calibration API**

- **Parameter spaces**: Continuous, discrete, categorical parameters  
The idea is to build on existing `Scenario` and `BatchRunner` concepts to form a unified experimentation workflow, where calibration, exploration, and optimization become different uses of the same execution infrastructure.
For example:
```python
ScenarioSpace = {
        "infection_rate": Continuous(0.01, 0.5),
        "mobility": Categorical(["low", "medium", "high"]),
        "seed": Integer(0, 1000)
}
```
Optimizers or samplers then draw concrete Scenario instances from this space.

Thus:

Scenario = one parameter realization  
ScenarioSpace = parameter domain


  - **Objective functions**: simple callables (single or multi-objective), with a consistent hook for replication aggregation; advanced multi-objective optimization logic stays in backends.

  - **Stopping conditions**: are primarily handled by backends (iteration, convergence, time budgets), while Mesa provides optional stop hooks, run-level timeouts, and checkpoint/resume support

  - **Results collection**: Structured logging with checkpointing

### 2. **Pluggable Optimization Backends**
Rather than implementing optimizers from scratch, provide 
a **general optimizer adapter interface**, and **one reference backend integration**, primarily to validate the API design.
A suitable candidate is **Optuna**, as it already supports:

- Bayesian and heuristic optimization,
- pruning and early stopping,
- parallel execution,
- robust experiment management.

Additionally, Mesa may include a simple built-in baseline sampler (e.g., random or grid search) for testing and small parameter spaces.

The purpose of these integrations is not to privilege specific tools, but to demonstrate how external libraries can drive Mesa experiments.

### 3. **Supporting Infrastructure**
To support experimentation workflows, several infrastructure pieces are useful:
- **Progress Monitoring**: Real-time visualization of optimization progress
- **Experiment Management**: Save/load calibration state, resume interrupted runs
- **Result Analysis**: Built-in tools for sensitivity analysis and parameter importance


### 4. **Optional Tensor-Friendly Output Interface**
Given the increasing use of spatial ABMs and ML workflows, it may be useful to make outputs easier to consume in tensor-based pipelines. However, to keep scope manageable, maybe only provide optional export helpers:

```Python
results.to_numpy()
results.to_tensor()  # optional dependency
```


## Some Design Questions

Some architectural questions worth discussing:

1. Should calibration be a new experiment type (like AnyLogic) or an extension to existing BatchRunner (like GAMA)?
2. How should experiment state and results be stored?
3. What standard patterns should exist for stochastic aggregation (mean, median, quantiles)?

## Next Steps

Would it make sense to:
1. Create a design RFC outlining the proposed API architecture?
2. Develop a minimal proof-of-concept with one backend (e.g., Optuna) to validate the design?


I'm happy to start with whichever approach the maintainers think would be most valuable. 
Looking forward to your thoughts!

**Comment by liyuezha** [comment]
Hi @quaquel, @EwoutH, thank you for the helpful discussion here. I’ve submitted my GSoC proposal based on this direction, titled "A Standard Experiment API for Calibration and Optimization in Mesa." The recent discussion around Scenario and experiment workflows was especially helpful in shaping and refining the proposal. I’d be very grateful for any feedback or comments!

**Reply by EwoutH** [reply]
Thanks for you proposal. Especially the diagram was very useful. A few questions/suggestions:

1. Scope: My main concern is whether all of this needs to be built inside Mesa. You propose quite a few new abstractions. Which of these do you think truly require Mesa internals, and which could just be a user writing a Python function or use Optuna, ema_workbench or another library? What's the minimum Mesa needs to own?
2. Starting from practice: Your timeline has the first real example in Week 9. What if you started the other way around: wire up a calibration of wolf-sheep with Optuna in plain Python first, and let the pain points tell you what abstractions to build?
5. If you had to cut scope: What's the smallest version of this project that would still be useful? What's the low-hanging fruit or underlying building block here?

**Reply by liyuezha** [reply]
Thanks a lot for the thoughtful feedback. This is very helpful!

### On scope, 
I fully agree that not everything I proposed needs to live inside Mesa. The pieces that seem most natural for Mesa to own are:
1. `ScnarioSpace`    
Mainly handling parameter validation, constraining parameter configurations, and instantiating scenarios.   I see this as a minimal layer to make parameter handling more consistent and reproducible, rather than a full search space abstraction.
2. `ExperimentRunner` and `ExperimentResult`
A lightweight execution layer together with a structured result object. This layer would be responsible for running the model for a given parameter configuration (including replications), collecting outputs (via DataCollector), and returning structured results (e.g., aggregated metrics, metadata) 
In this design, `Experiment` would likely be just a thin convenience wrapper, rather than the center of a large abstraction stack.




### For the first example,
I really like the suggestion to start with practice first. That makes a lot of sense, and I think that is a better way to structure the work.
For the first example, I would like to **use Optuna to calibrate a simple Schelling model**. Compared to wolf-sheep, the summary statistics are more straightforward, and the calibration target is easier to explain. This makes it a clearer starting point for identifying the minimal interface. So the revised plan would be:
1. Build an end-to-end calibration example first in plain Python. Using a Schelling model, without introducing new Mesa abstractions prematurely. 
2. Use that workflow to identify repeated friction points, such as parameter space definition, repeated model execution with replications, structured result collection......
3. Only then formalize the minimal abstractions that seem justified
 
Instead of waiting until Week 9 for the first real example, I would move the example pipeline to the beginning **community bounding phase** and treat it as the design driver for the API. I think this would lead to a much more grounded proposal and help ensure that Mesa only owns the pieces that are actually painful to reproduce in every workflow.


### On minimum version,
If I had to cut the scope to the smallest useful version, I would reduce it to a single building block: **A standardized workflow to evaluate a parameterized Mesa model and return structured results**, which would include:

1. The minimal abstractions for a standardized evaluation interface, likely centered on a minimal ExperimentRunner and ScenarioSpace
2. One end-to-end example integrating Mesa with Optuna, without introducing a heavy abstraction layer

This would already provide value by making calibration workflows reproducible, reducing boilerplate code for repeated experiments, and demonstrating best practices for integrating Mesa with external tools. Additional abstractions (such as a full Experiment API or more advanced components) could then be introduced incrementally if they prove necessary in practice.
